firewood/src/root_store.rs

```diff
@@ -57,24 +75,126 @@ impl RootStore {
         Ok(())
     }
 
-    /// `get` returns the address of a revision.
+    /// `get` retrieves a committed revision by its hash.
+    ///
+    /// To retrieve a committed revision involves a few steps:
+    /// 1. Check if the committed revision is cached.
+    /// 2. If the committed revision is not cached, query the underlying
+    ///    datastore for the revision's root address.
+    /// 3. Construct the committed revision.
     ///
     /// Args:
     /// - hash: the hash of the revision
     ///
     /// # Errors
     ///
-    ///  Will return an error if unable to query the underlying datastore.
+    ///  Will return an error if unable to query the underlying datastore or if
+    ///  the stored address is invalid.
+    ///
+    /// # Panics
+    ///
+    ///  Will panic if the latest revision does not exist.
     pub fn get(
         &self,
         hash: &TrieHash,
-    ) -> Result<Option<LinearAddress>, Box<dyn std::error::Error + Send + Sync>> {
+    ) -> Result<Option<CommittedRevision>, Box<dyn std::error::Error + Send + Sync>> {
+        // 1. Check if the committed revision is cached.
+        if let Some(v) = self.revision_cache.lock().get(hash) {
+            return Ok(Some(v));
+        }
+
+        // 2. If the committed revision is not cached, query the underlying
+        //    datastore for the revision's root address.
         let Some(v) = self.items.get(**hash)? else {
             return Ok(None);
         };
 
         let array: [u8; 8] = v.as_ref().try_into()?;
+        let addr = LinearAddress::new(u64::from_be_bytes(array))
```

Comment:
This code seems racy. There's nothing to prevent two threads from calling get() on the same hash and ending up here:

Thread 1 runs to this point and releases the revision_cache lock, not seeing the hash it's looking for
Thread 2 also runs to this point and releases the revision_cache lock, again not seeing the hash it's looking for

I think you'll need to hold the revision_cache lock while doing all this other work. I think the mutex will only be locked while querying the underlying store, which should be pretty fast, so I think we should keep the lock around.